go/adbc/sqldriver/driver.go

```diff
@@ -0,0 +1,695 @@
+	return c.PrepareContext(context.Background(), query)
+}
+
+// PrepareContext returns a prepared statement, bound to this connection.
+// Context is for the preparation of the statement. The statement must not
+// store the context within the statement itself.
+func (c *conn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	opts := GetOptionsFromCtx(ctx)
+	s, err := c.Conn.NewStatement(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.SetSqlQuery(query); err != nil {
+		s.Close()
+		return nil, err
+	}
+
+	paramSchema, err := s.GetParameterSchema()
+	var adbcErr adbc.Error
+	if errors.As(err, &adbcErr) {
+		if adbcErr.Code != adbc.StatusNotImplemented {
+			return nil, err
+		}
+	}
+
+	return &stmt{stmt: s, paramSchema: paramSchema}, nil
+}
+
+type tx struct {
+	ctx  context.Context
+	conn adbc.Connection
+}
+
+func (t tx) Commit() error {
+	if err := t.conn.Commit(t.ctx); err != nil {
+		return err
+	}
+
+	return t.conn.(adbc.PostInitOptions).SetOption(adbc.OptionKeyAutoCommit, adbc.OptionValueEnabled)
+}
+
+func (t tx) Rollback() error {
+	if err := t.conn.Rollback(t.ctx); err != nil {
+		return err
+	}
+	return t.conn.(adbc.PostInitOptions).SetOption(adbc.OptionKeyAutoCommit, adbc.OptionValueEnabled)
+}
+
+type stmt struct {
+	stmt        adbc.Statement
+	paramSchema *arrow.Schema
+}
+
+func (s *stmt) Close() error {
+	return s.stmt.Close()
+}
+
+func (s *stmt) NumInput() int {
+	if s.paramSchema == nil {
+		return -1
+	}
+
+	return len(s.paramSchema.Fields())
+}
+
+func (s *stmt) Exec(args []driver.Value) (driver.Result, error) {
+	return nil, driver.ErrSkip
+}
+
+func (s *stmt) Query(args []driver.Value) (driver.Rows, error) {
+	return nil, driver.ErrSkip
+}
+
+func checkType[T any](val any) bool {
+	switch val.(type) {
+	case T, *T:
+	default:
+		return false
+	}
+	return true
+}
+
+func isCorrectParamType(typ arrow.Type, val driver.Value) bool {
+	switch typ {
+	case arrow.BINARY:
+		return checkType[[]byte](val)
+	case arrow.BOOL:
+		return checkType[bool](val)
+	case arrow.INT8:
+		return checkType[int8](val)
+	case arrow.UINT8:
+		return checkType[uint8](val)
+	case arrow.INT16:
+		return checkType[int16](val)
+	case arrow.UINT16:
+		return checkType[uint16](val)
+	case arrow.INT32:
+		return checkType[int32](val)
+	case arrow.UINT32:
+		return checkType[uint32](val)
+	case arrow.INT64:
+		return checkType[int64](val)
+	case arrow.UINT64:
+		return checkType[uint64](val)
+	case arrow.STRING:
+		return checkType[string](val)
+	case arrow.FLOAT32:
+		return checkType[float32](val)
+	case arrow.FLOAT64:
+		return checkType[float64](val)
+	case arrow.DATE32:
+		return checkType[arrow.Date32](val)
+	case arrow.DATE64:
+		return checkType[arrow.Date64](val)
+	case arrow.TIME32:
+		return checkType[arrow.Time32](val)
+	case arrow.TIME64:
+		return checkType[arrow.Time64](val)
+	case arrow.TIMESTAMP:
+		return checkType[arrow.Timestamp](val)
+	}
+	// TODO: add more types here
+	return true
```

Comment:
In this situation, I'm having the default be to let the underlying driver decide whether or not the parameter is of the wrong type rather than making that decision pre-emptively here. (for example, if the driver did not know what the type was and provided a `NullType` then we just let anything through). At least that was the intent, as we add more types into this check, we'd be providing better type-checking rather than *enabling* types that previously didn't work at all.

Reply:
Ah, got it - thanks for the explanation.